packages/core/src/VDocument.ts

```diff
@@ -152,4 +153,199 @@ export class VDocument {
+                duplicatedNodes.push(ancestor.splitAt(child));
+                child = ancestor;
+                return ancestor === firstList;
+            });
+            if (!firstList.hasChildren()) {
+                firstList.remove();
+            }
+        }
+        const last = nodes[nodes.length - 1];
+        const lastList = this._listAncestor(last);
+        if (lastList) {
+            let child = last;
+            last.ancestor(ancestor => {
+                duplicatedNodes.push(ancestor.splitAt(child));
+                child = ancestor;
+                return ancestor === lastList;
+            });
+            if (!lastList.hasChildren()) {
+                lastList.remove();
+            }
+        }
+
+        // Create and insert the list before the common ancestor to the first
+        // last node, unless the first node is within a list, in which case
+        // insert before the list instead to avoid inserting the new list within
+        // a list.
+        let list = new ListNode(htmlTag);
+        const listAncestor = this._listAncestor(nodes[0]);
+        if (listAncestor) {
+            listAncestor.before(list);
+        } else {
+            const commonAncestor = nodes[0].ancestor(ancestor => {
+                return !!nodes[nodes.length - 1].ancestor(
+                    otherAncestor => otherAncestor === ancestor,
+                );
+            });
+            commonAncestor.insertBefore(
+                list,
+                nodes[0].ancestor(ancestor => ancestor.parent === commonAncestor) || nodes[0],
+            );
+        }
+
+        // Move the nodes to the list
+        withMarkers(() => {
+            // Remove the children of nested lists (they will be moved together
+            // with their parent) and convert nested lists to the new type.
+            const nestedLists = nodes.filter(this._isNestedList.bind(this));
+            nodes = nodes.filter(node => {
+                return !node.ancestor(ancestor => nestedLists.includes(ancestor));
+            });
+            nestedLists.forEach(nestedList => ((nestedList as ListNode).htmlTag = htmlTag));
+            // Move the nodes.
+            nodes.forEach(node => {
+                list.append(node);
+            });
+            // Remove empty lists.
+            duplicatedNodes.concat(nodes).forEach(node => {
+                if (this._isList(node) && !node.hasChildren()) {
+                    node.remove();
+                }
+            });
+        });
+
+        // If the new list is after or before a list of the same type, merge
+        // them (eg: <ol><li>a</li></ol><ol><li>b</li></ol> =>
+        // <ol><li>a</li><li>b</li></ol>).
+        const previousSibling = list.previousSibling();
+        if (this._isList(previousSibling) && (previousSibling as ListNode).htmlTag === htmlTag) {
+            list.children.slice().forEach(child => previousSibling.append(child));
+            list.remove();
```

Comment:
I'm wondering if we might get rid of `withMarkers` by calling `mergeWith` enough time to empty the list (since `mergeWith` uses `withMarkers` internally.)

Reply:
Good question. Is it really worth investigating though?

Reply:
No, I agree that iti s out of scope.